pkg/clusterapi/aws/actuator_test.go

```diff
@@ -0,0 +1,338 @@
+		Filters: []*ec2.Filter{{Name: aws.String("tag:Name"), Values: []*string{&vpcName}}},
+	}
+	describeVpcsOutput := ec2.DescribeVpcsOutput{
+		Vpcs: []*ec2.Vpc{
+			{
+				VpcId: aws.String(vpcID),
+			},
+		},
+	}
+	mockAWSClient.EXPECT().DescribeVpcs(&describeVpcsInput).Return(&describeVpcsOutput, nil)
+}
+
+func addDescribeSubnetsMock(mockAWSClient *MockAWSClient, az, vpcID, subnetID string) {
+	input := ec2.DescribeSubnetsInput{
+		Filters: []*ec2.Filter{
+			{Name: aws.String("vpc-id"), Values: []*string{aws.String(vpcID)}},
+			{Name: aws.String("availability-zone"), Values: []*string{aws.String(az)}},
+		},
+	}
+	output := ec2.DescribeSubnetsOutput{
+		Subnets: []*ec2.Subnet{
+			{
+				SubnetId: aws.String(subnetID),
+			},
+		},
+	}
+	mockAWSClient.EXPECT().DescribeSubnets(&input).Return(&output, nil)
+}
+
+func addDescribeSecurityGroupsMock(mockAWSClient *MockAWSClient, vpcID, vpcName string, isMaster, isInfra bool) {
+	// Using the input builder from the production code. Behavior of this function is tested separately.
+	input := buildDescribeSecurityGroupsInput(vpcID, vpcName, isMaster, isInfra)
+	output := ec2.DescribeSecurityGroupsOutput{
+		SecurityGroups: []*ec2.SecurityGroup{},
+	}
+	// Using the explicit index where we store our name filters, may want to search for it if this
+	// proves too frail.
+	// Add a fake ID for each security group we're searching for.
+	for _, groupName := range input.Filters[1].Values {
+		output.SecurityGroups = append(output.SecurityGroups, &ec2.SecurityGroup{GroupId: aws.String(fmt.Sprintf("%s-ID", *groupName))})
+	}
+	mockAWSClient.EXPECT().DescribeSecurityGroups(input).Return(&output, nil)
+}
+
+// testClusterDeployment creates a new test ClusterDeployment
+func testClusterDeployment() *clustopv1.ClusterDeployment {
+	clusterDeployment := &clustopv1.ClusterDeployment{
+		ObjectMeta: metav1.ObjectMeta{
+			UID:       testClusterDeploymentUUID,
+			Name:      testClusterDeploymentName,
+			Namespace: testNamespace,
+		},
+		Spec: clustopv1.ClusterDeploymentSpec{
+			MachineSets: []clustopv1.ClusterMachineSet{
+				{
+					ShortName: "",
+					MachineSetConfig: clustopv1.MachineSetConfig{
+						Infra:    true,
+						Size:     1,
+						NodeType: clustopv1.NodeTypeMaster,
+					},
+				},
+				{
+					ShortName: "compute",
+					MachineSetConfig: clustopv1.MachineSetConfig{
+						Infra:    false,
+						Size:     1,
+						NodeType: clustopv1.NodeTypeCompute,
+					},
+				},
+			},
+			Hardware: clustopv1.ClusterHardwareSpec{
+				AWS: &clustopv1.AWSClusterSpec{
+					SSHUser:     "clusteroperator",
+					Region:      testRegion,
+					KeyPairName: "libra",
+				},
+			},
+			ClusterVersionRef: clustopv1.ClusterVersionReference{
+				Name:      testClusterVerName,
+				Namespace: testClusterVerNS,
+			},
+		},
+	}
+	return clusterDeployment
+}
+
+func testCluster() *capiv1.Cluster {
+	clusterDeployment := testClusterDeployment()
+	cluster, _ := clusterdeployment.BuildCluster(clusterDeployment)
```

Comment:
I don't like creating a cluster indirectly from a clusterdeployment for these tests. I don't think that it's any more difficult to create the cluster directly, and then it removes a possible layer of complication when analysing why a test is failing. In the end, I expect that the `ProviderConfig` in cluster is going to be pretty lean, where the test will be filling out a lot more in clusterdeployment than it would if creating a cluster directly.

If you are going to use `BuildCluster`, move it to a more centrally accessible location rather than leaving it in the package of the clusterdeployment controller.

Reply:
I'm a bit torn on this one, I considered it at the time and just tried doing up a patch for it but it just doesn't feel right, we need a ClusterDeployment for the ProviderConfig right now either way, it feels wrong to duplicate what BuildCluster does and introduce more room for deviations when the logic is right there and easily accessible. We can evolve what it means to convert a ClusterDeployment to a capi.Cluster over time with less mass test updates. I can see how it's another thing to debug if something goes wrong, but with the requirement on a ClusterDeployment for the provider config anyhow right now, this seems like the lesser to me. If this is reasonable I'd like to keep re-using, and more the function somewhere more suitable.

Reply:
That's fine. I would like in the very near future to create separate types for the ProviderConfig in Cluster and Machine. My concern is that if this test is using `BuildCluster` now, then it always will, even after Cluster ProviderConfig is no longer taking a ClusterDeploymentSpec. But that's not terrible anyway.